File: erlsqlacore.erl
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% module: erlsqlacore
%% description: use erlport to connect Erlang processes to SQLAlchemy 
%% author: github.com/snwight
%% license: dbad
%% date: oct 2012
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
-module(erlsqlacore).
-export([init/0, start/2, get/2, upsert/2, remove/2]).

init() ->
    %% start up python sibling module - it will await our instructions
    Port = open_port({spawn, "python -u ErlSqlaCore.py"},
		     [{packet, 1}, binary, 
		      {env, [{"PYTHONPATH", "../src"}]}]),
    port_command(Port, term_to_binary({port, term_to_binary(Port)})),
    handle(Port).

handle(Port) ->
    receive
        {Port, {data, Data}} ->
	    case binary_to_term(Data) of 
		<<"ported">> -> Port;
		<<"started">> -> "started up";
		X -> X
	    end;
	_Other ->
	    port_close(Port),
	    {whoops, _Other}
    after
        5000 ->
            {error, timeout}
    end.

start(Port, ConfigString) ->
    port_command(Port, term_to_binary({start, term_to_binary(ConfigString)})),
    handle(Port).

get(Port, ArgList) ->
    port_command(Port, term_to_binary({get, term_to_binary(ArgList)})),
    handle(Port).

upsert(Port, ArgList) ->
    port_command(Port, term_to_binary({upsert, term_to_binary(ArgList)})),
    handle(Port).

remove(Port, ArgList) ->
    port_command(Port, term_to_binary({remove, term_to_binary(ArgList)})),
    handle(Port).

    

File: ErlSqlaCore.py
###############################################################################
# module: ErlSqlaCore.py
# description: Python sibling of erlsqlacore.erl, handles message passing 
# between Erlang node/s and SQLACore module (simple SQLAlchemy Core api api)
# ...fully dependent upon erlport, by github.com/hdima/erlport
# author: snwight, northwight@gmail.com
# repos: github.com/snwight/erlsqlacore
# license: dbad
# date: oct 2012
###############################################################################
from SqlaCore import *
from erlport import Port, Protocol
from erlport import Atom, String

class ErlSqlaCore(Protocol):
    def __init__(self):
        pass

    def handle_port(self, port=None):
        # store active port descriptor
        self.activePort = port
        return "ported"

    def handle_start(self, configString=None):
        # crank up SQLAlchemy engine
        #        self.sqlaconfig = String(configString)
        self.sqlaconfig = "sqlite:////tmp/test.db"
        self.sqlacore = SqlaCore(self.sqlaconfig)
        if self.sqlacore:
            self.sqlacore.loadSchema("/Users/snwight/erlsqlacore/schema1.sql")
        return "started"
        
    def handle_get(self, argList=None):
        rowSet = self.sqlacore.get(argList)
        # return FULL rowset result, for now!
        self.activePort.write(rowSet)
        return rowSet

    def handle_upsert(self, argList=None):
        rowCount = self.sqlacore.upsert(argList)
        self.activePort.write(rowCount)
        return rowCount

    def handle_remove(self, argList=None):
        rowCount = self.sqlacore.remove(argList)
        self.activePort.write(rowCount)
        return rowCount

if __name__ == "__main__":
    proto = ErlSqlaCore()
    proto.run(Port(use_stdio=True))


